dali/pipeline/operator/builtin/external_source_test.cc

```diff
@@ -22,9 +24,189 @@
+ protected:
+  static constexpr bool is_cpu = std::is_same_v<Backend, CPUBackend>;
+
+  void CreateAndSerializePipeline(bool repeat_last) {
+    auto pipeline = std::make_unique<Pipeline>(batch_size_, num_threads_, device_id_, false, 1,
+                                               false);
+    pipeline->AddOperator(
+            OpSpec("ExternalSource")
+                    .AddArg("repeat_last", repeat_last)
+                    .AddArg("device", std::string{is_cpu ? "cpu" : "gpu"})
+                    .AddArg("name", input_name_)
+                    .AddOutput(input_name_, std::string{is_cpu ? "cpu" : "gpu"}),
+            input_name_);
+
+    std::vector<std::pair<std::string, std::string>> outputs = {
+            {input_name_, is_cpu ? "cpu" : "gpu"},
+    };
+
+    pipeline->SetOutputDescs(outputs);
+
+    serialized_pipeline_ = pipeline->SerializeToProtobuf();
+  }
+
+
+  void FeedExternalInput(daliPipelineHandle *h) {
+    daliSetExternalInputBatchSize(h, input_name_.c_str(), batch_size_);
+    std::vector<int64_t> shapes;
+    std::string data;
+    for (int i = 0; i < batch_size_; i++) {
+      auto sample = LoadTestFile(test_files_[i % test_files_.size()]);
+      shapes.push_back(static_cast<int64_t>(sample.length()));
+      data += sample;
+    }
+    if constexpr (is_cpu) {
+      daliSetExternalInput(h, input_name_.c_str(), device_type_t::CPU, data.data(),
+                           dali_data_type_t::DALI_UINT8, shapes.data(), 1, nullptr,
+                           DALI_ext_force_copy);
+    } else {
+      thrust::host_vector<uint8_t> data_cpu{data.begin(), data.end()};
+      thrust::device_vector<uint8_t> data_gpu = data_cpu;
+      daliSetExternalInput(h, input_name_.c_str(), device_type_t::GPU,
+                           thrust::raw_pointer_cast(data_gpu.data()),
+                           dali_data_type_t::DALI_UINT8, shapes.data(), 1, nullptr,
+                           DALI_ext_force_copy);
+    }
+  }
+
+
+  /**
+   * Check, if the "depleted" trace exists.
+   */
+  void DoesDepletedTraceExist(daliPipelineHandle *h) {
+    // The "depleted" trace should always exist.
+    ASSERT_TRUE(daliHasOperatorTrace(h, input_name_.c_str(), depleted_trace_name_.c_str()));
+  }
+
+
+  /**
+   * Verify the value of "depleted" trace.
+   * @param shall_be_depleted Expected value.
+   */
+  void IsDepletedTraceCorrect(daliPipelineHandle *h, bool shall_be_depleted) {
+    EXPECT_STREQ(daliGetOperatorTrace(h, input_name_.c_str(), depleted_trace_name_.c_str()),
+                 shall_be_depleted ? "true" : "false");
+  }
+
+
+  const int batch_size_ = 3;
+  const int num_threads_ = 2;
+  const int device_id_ = 0;
+  daliPipelineHandle handle_;
+  std::string serialized_pipeline_;
+  const std::string depleted_trace_name_ = "depleted";
+  const std::string input_name_ = "INPUT";
+  const std::vector<std::string> test_files_ = {
+          make_string(testing::dali_extra_path(), "/db/single/jpeg/100/swan-3584559_640.jpg"),
```

Comment:
If you just want to check the operator traces, isn't reading a file an overkill? You could just feed a few bytes of hard-coded data - it would be easier to verify, too.

Reply:
From my perspective, this would rather be equivalent. Both are deterministic and both would require approximately the same amount of lines of code. @jantonguirao , what's your opinion?

Reply:
Hand feeding few bytes would be preferable if we care about testing time. No strong push from my end though.

Reply:
👍🏻 , changed to 10 bytes per sample.